qualweb/qualweb
# @qualweb/best-practices: fix unit tests

This PR fixes up the unit tests for @qualweb/best-practices (issue #4). If you decide to merge it, I recommend you *squash* and merge instead of retaining the history. There's several intermediate commits that I don't think would be useful in the commit history.

A brief overview of the changes follows.

# Test template

I added a `buildTest()` "test builder" that produces a test suite for a given BP rule. Looking over the test cases, they were virtually identical (barring a bit of code style or change of import style). In order, they would: 

1. navigate to a URL (or load its HTML) using @qualweb/dom (in some cases, using a non-existent function `getDom()`)
2. inject the requisite BP scripts (with dependencies)
3. run the specific BP rule on the document
4. check the outcome

Wrapping that together in a template like `buildTest()` seemed smart - DRY. If a test should require more granular execution in the future, you can still just write a full suite for it.

Additionally, only a scant few tests actually ran in the state they were initially. They were using non-existent functions or defining a "require" variable that clashed with the `require` keyword/function in NodeJS. These changes indirectly address that, too.

# Fixtures

Most of the tests referenced remote URLs as the documents to evaluate against. Some (BP19 and forward) defined them in JS files as HTML code. Using a remote server for the documents is a bad idea, and keeping HTML in JS files instead of just HTML files is harder to maintain (and read).

I've opted to create separate a separate fixture folder for each of the test cases under "test/fixtures/testcases". So you have "test/fixtures/testcases/BP2", "test/fixtures/testcases/BP4", "test/fixtures/testcases/BP17", etc. In each folder, I've downloaded the appropriate testcases from their remote URLs. For the JS files, I've taken out the HTML code and placed them in matching HTML files.

For each of these folders, I added a "testcases.json" file, that contains an array of cases to test, each one with an outcome and pointing to its matching HTML file.

The result is that all the unit tests are fully unified in structure, and can run entirely offline, removing the risk of false positive for errors if the remote server should fail to respond (in at least two cases, it would likely have failed, since HTTPS connections to it were being rejected).

# Minor changes / other

- As part of downloading local versions of the test cases, I've added a W3C logo that was referenced by BP13. Reading guidance on https://www.w3.org/policies/logos , I think adding it for this use is acceptable.
- url.spec.mjs was verbose and wasn't using proper setup/teardown for puppeteer. This has been fixed.
- Two rules no longer have an implementation, and would fail their test: BP3 and BP14. I've retained them in the list of tests for now, but if the rules have been removed, the tests should probably also just be deleted.
- Two rules have no tests defined: BP17 and BP18. I've added "pending" tests, so that they will show up when running mocha, but otherwise not do anything.
- Most of the test files were using CRLF line endings (Windows-style). *All* other files in the monorepo use LF in the index. For consistency's sake I've changed the CRLF files to LF. I don't expect anyone to notice the change.
